### PR TITLE
cluster: Manage Pod Disruption Budget (#214)

### LINKED ISSUE #214: Manage Pod Disruption Budget
* Feature Request

Operator should manage PDB to ensure high availability and protect Scylla from voluntary disruptions.  
https://kubernetes.io/docs/concepts/workloads/pods/disruptions/

### PR DESCRIPTION
**Pull request by zimnx**
A PDB with maxUnavailability of 1 is now managed by the Operator.

Fixes #214

### CODE REVIEW

File: pkg/util/apply/pdb.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	"k8s.io/api/policy/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/record"
+)
+
+// PodDisruptionBudget creates or updates PDB. If desired object is equal (based on calculated hash)
+// to existing one update is not called.
+func PodDisruptionBudget(ctx context.Context, recorder record.EventRecorder, client kubernetes.Interface, pdb *v1beta1.PodDisruptionBudget) (*v1beta1.PodDisruptionBudget, error) {
```
Comment (tnozicka) [inline pkg/util/apply/pdb.go:19]: Would you mind wiring it using listers (caches)? No expectations, just easy erroring out until informers catch up.

File: pkg/util/apply/pdb_test.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/scylladb/scylla-operator/pkg/naming"
+	"github.com/scylladb/scylla-operator/pkg/util/apply"
+	"k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/record"
+)
+
+func TestApplyPodDisruptionBudget(t *testing.T) {
```
Comment (tnozicka) [inline pkg/util/apply/pdb_test.go:18]: would you mind making it a table based test?

File: pkg/util/apply/events.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		kinds, _, _ := scheme.Scheme.ObjectKinds(obj)
+		if len(kinds) > 0 && !kinds[0].Empty() {
+			gvk = kinds[0]
+		} else {
+			gvk = schema.GroupVersionKind{Group: unknown, Version: unknown, Kind: unknown}
+		}
+	}
+
+	var name, namespace string
+	accessor, err := meta.Accessor(obj)
```
Comment (tnozicka) [inline pkg/util/apply/events.go:29]: these aren't really properties of the type (by the func name), maybe a separate function?

File: pkg/controllers/cluster/resource/resource.go [hunk 4]
```diff
@@ -474,6 +475,21 @@ func agentContainer(r scyllav1.RackSpec, c *scyllav1.ScyllaCluster) corev1.Conta
 	return cnt
 }
 
+func PodDisruptionBudget(c *scyllav1.ScyllaCluster) *v1beta1.PodDisruptionBudget {
```
Comment (tnozicka) [inline pkg/controllers/cluster/resource/resource.go:478]: nit: getPodDisruptionBudget (or makePodDisruptionBudget)

File: pkg/util/apply/pdb.go (hunk 1 above)
Comment (tnozicka) [reply pkg/util/apply/pdb.go:19]: It should also merge metadata to ensure it has e.g. correct labels

[the remark this answers is not in the document]
Comment (zimnx) [reply pkg/util/apply/pdb.go:19]: There's already an `apply` in package name. On caller side, it looks like `apply.PodDisruptionBudget`

File: pkg/util/apply/hash_test.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+type A struct {
+	I int
+	S string
+}
+
+func TestDeepHashObject(t *testing.T) {
+	tests := []func() interface{}{
+		func() interface{} { return 123 },
```
[the remark this answers is not in the document]
Comment (zimnx) [reply pkg/util/apply/hash_test.go:16]: By having a function it ensures that for each call, different (but equal) object is returned, and this test validates whether hashes of different equal object are the same.

File: pkg/util/apply/pdb.go (hunk 1 above)
Comment (tnozicka) [reply pkg/util/apply/pdb.go:19]: actually, hashing the whole object is sufficient for PDB

File: pkg/util/apply/hash.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/scylladb/scylla-operator/pkg/naming"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func deepHashObject(obj interface{}) string {
+	hasher := sha256.New()
+	_ = json.NewEncoder(hasher).Encode(obj)
+	return hex.EncodeToString(hasher.Sum(nil)[0:])
+}
```
[the remark this answers is not in the document]
Comment (zimnx) [reply pkg/util/apply/hash.go]: Added multiple objects parameter, but left json.Encoder. It's basically doing the same thing.

File: pkg/controllers/cluster/sync.go [hunk 7]
```diff
@@ -53,6 +54,12 @@ func (cc *ClusterReconciler) sync(c *scyllav1.ScyllaCluster) error {
 		return errors.Wrap(err, "failed to sync headless service")
 	}
 
+	// Sync Cluster Pod Disruption Budget
+	if err := cc.syncPodDisruptionBudget(ctx, c); err != nil {
+		cc.Recorder.Event(c, corev1.EventTypeWarning, naming.ErrSyncFailed, MessagePodDisruptionBudgetSyncFailed)
```
Comment (tnozicka) [inline pkg/controllers/cluster/sync.go:59]: preexisting pattern (but sync failures (retriable) don't usually manifest into events, degraded condition on the status would be more fitting I guess)

File: pkg/naming/names.go [hunk 8]
```diff
@@ -51,6 +51,10 @@ func HeadlessServiceNameForCluster(c *scyllav1.ScyllaCluster) string {
 	return fmt.Sprintf("%s-client", c.Name)
 }
 
+func PodDisruptionBudgetName(c *scyllav1.ScyllaCluster) string {
+	return fmt.Sprintf("%s", c.Name)
```
Comment (tnozicka) [inline pkg/naming/names.go:55]: this looks looks noop to me

File: pkg/util/apply/hash.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func HashObjects(objs ...interface{}) string {
+	hasher := sha256.New()
+	for _, obj := range objs {
+		_ = json.NewEncoder(hasher).Encode(obj)
+	}
+	return hex.EncodeToString(hasher.Sum(nil))
```
Comment (tnozicka) [inline pkg/util/apply/hash.go:20]: base64.StdEncoding.EncodeToString?

File: pkg/util/apply/hash.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+	"encoding/hex"
+	"encoding/json"
+
+	"github.com/scylladb/scylla-operator/pkg/naming"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func HashObjects(objs ...interface{}) string {
+	hasher := sha256.New()
```
Comment (tnozicka) [inline pkg/util/apply/hash.go:16]: I'd go with 512

File: pkg/util/apply/hash.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+	for _, obj := range objs {
+		_ = json.NewEncoder(hasher).Encode(obj)
```
Comment (tnozicka) [inline pkg/util/apply/hash.go:18]: error handling

File: pkg/util/apply/hash.go [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+	}
+	return hex.EncodeToString(hasher.Sum(nil))
+}
+
+func setHash(obj runtime.Object) error {
```
Comment (tnozicka) [inline pkg/util/apply/hash.go:23]: would it be better to accept ObjectMeta directly?

**Review by tnozicka — COMMENTED** [review]
just flew over it, gonna give it a second pass when the base PR merges

File: pkg/controllers/cluster/sync.go (hunk 7 above)
Comment (zimnx) [reply pkg/controllers/cluster/sync.go:59]: Lets fix it once for all failures as a follow up in 1.2: https://github.com/scylladb/scylla-operator/issues/513

**Comment by tnozicka** [comment]
I've written a similar apply package in https://github.com/scylladb/scylla-operator/tree/b8a6db3cd9337002791faa0d7426555a74109683/pkg/resourceapply if you want to have a look and unify our methods to have a best of boths.

File: test/e2e/set/scyllacluster/scyllacluster_evictions.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+		verifyScyllaCluster(ctx, f.KubeClient(), sc)
+
+		framework.By("Should allow eviction of first pod")
+		e := &policyv1beta1.Eviction{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      getNodeName(0, sc),
+				Namespace: f.Namespace(),
+			},
+		}
+		err = f.KubeClient().CoreV1().Pods(f.Namespace()).Evict(ctx, e)
```
Comment (tnozicka) [inline test/e2e/set/scyllacluster/scyllacluster_evictions.go:94]: using KubeAdminClient here would safe a lot of lines setting up the individual role

File: test/e2e/set/scyllacluster/helpers.go [hunk 14]
```diff
@@ -196,10 +196,15 @@ func getScyllaClient(ctx context.Context, client corev1client.CoreV1Interface, s
 }
 
 func getFirstNonSeedNodeName(sc *scyllav1.ScyllaCluster) string {
+	return getNodeName(2, sc)
+}
+
+func getNodeName(idx int, sc *scyllav1.ScyllaCluster) string {
```
Comment (tnozicka) [inline test/e2e/set/scyllacluster/helpers.go:202]: nit: I'd reverse the arg order for readability but that may be opinionated

File: test/e2e/set/scyllacluster/scyllacluster_evictions.go [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+		framework.By("Waiting for the ScyllaCluster to deploy")
+		waitCtx1, waitCtx1Cancel := contextForRollout(ctx, sc)
+		defer waitCtx1Cancel()
+		sc, err = waitForScyllaClusterState(waitCtx1, f.ScyllaClient().ScyllaV1(), sc.Namespace, sc.Name, scyllaClusterRolledOut)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		verifyScyllaCluster(ctx, f.KubeClient(), sc)
+
+		framework.By("Should allow eviction of first pod")
```
Comment (tnozicka) [inline test/e2e/set/scyllacluster/scyllacluster_evictions.go:87]:
```suggestion
		framework.By("Allowing the first pod to be evicted")
```

File: test/e2e/set/scyllacluster/scyllacluster_evictions.go [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+		e := &policyv1beta1.Eviction{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      getNodeName(0, sc),
+				Namespace: f.Namespace(),
+			},
+		}
+		err = f.KubeClient().CoreV1().Pods(f.Namespace()).Evict(ctx, e)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		framework.By("Should forbid eviction of second pod")
```
Comment (tnozicka) [inline test/e2e/set/scyllacluster/scyllacluster_evictions.go:97]: same here, the ginkgo statements should be readable as a sentence - so you have:
`It("should allow one disruption")`
`By("Should forbid eviction of second pod")`
vs.
`It("should allow one disruption")`
`By("Refusing to evict a second pod")`

File: test/e2e/set/scyllacluster/scyllacluster_evictions.go [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+		e = &policyv1beta1.Eviction{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      getNodeName(1, sc),
+				Namespace: f.Namespace(),
+			},
+		}
+		err = f.KubeClient().CoreV1().Pods(f.Namespace()).Evict(ctx, e)
+		o.Expect(err).To(o.HaveOccurred())
```
Comment (tnozicka) [inline test/e2e/set/scyllacluster/scyllacluster_evictions.go:105]: guess this is redundant given the next check

File: test/e2e/set/scyllacluster/scyllacluster_evictions.go [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+		sc, err = f.ScyllaClient().ScyllaV1().ScyllaClusters(f.Namespace()).Create(ctx, sc, metav1.CreateOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		framework.By("Waiting for the ScyllaCluster to deploy")
+		waitCtx1, waitCtx1Cancel := contextForRollout(ctx, sc)
+		defer waitCtx1Cancel()
+		sc, err = waitForScyllaClusterState(waitCtx1, f.ScyllaClient().ScyllaV1(), sc.Namespace, sc.Name, scyllaClusterRolledOut)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		verifyScyllaCluster(ctx, f.KubeClient(), sc)
```
Comment (tnozicka) [inline test/e2e/set/scyllacluster/scyllacluster_evictions.go:85]: extend the verifyScyllaCluster to check the presence of the PDB object and it's values

**Review by tnozicka — COMMENTED** [review]
Thanks for the update! We can adjust some minor things over time but the e2e may be easy enough to address before merge.

**Review by tnozicka — APPROVED** [review]
/lgtm
thanks!

**Pull request merged**